pkg/export/gcm/promtest/promtest.go

```diff
@@ -0,0 +1,316 @@
+		// to -8m rather.
+		currTime: time.Now().Add(-1 * time.Hour),
+
+		backends:               map[string]backend{},
+		expectationsPerBackend: map[string]model.Matrix{},
+	}
+
+	for _, b := range backends {
+		api, extLset := b.start(t, e)
+		it.backends[b.Ref()] = backend{b: b, api: api, extLset: extLset}
+	}
+	if len(it.backends) == 0 {
+		t.Fatal("no backends specified, at least has to be passed e.g. promtest.Prometheus or promtest.LocalExportWithGCM")
+	}
+
+	return it
+}
+
+type Backend interface {
+	Ref() string
+	start(t testing.TB, env e2e.Environment) (api v1.API, extraLset map[string]string)
+	injectScrapes(t testing.TB, scrapeRecordings [][]*dto.MetricFamily, timeout time.Duration)
+}
+
+type ExpectationsRecorder interface {
+	Expect(val float64, metric prometheus.Metric, b Backend) ExpectationsRecorder
+
+	// TODO(bwplotka): Add histogram support.
+}
+
+type Scrape func(after time.Duration) ExpectationsRecorder
+
+// RecordScrapes allows recording SDK and expected samples for interesting cases.
+func (it *ingestionTest) RecordScrapes(recordingFunc func(r prometheus.Registerer, scrape Scrape)) {
+	fmt.Printf("%s Recording scrapes for test=%q\n", it.t.Name(), it.testID)
+	r := prometheus.NewRegistry()
+
+	var scrapeRecordings [][]*dto.MetricFamily
+
+	recordingFunc(prometheus.WrapRegistererWith(map[string]string{"test": it.testID}, r), func(after time.Duration) ExpectationsRecorder {
+		it.t.Helper()
+
+		if after <= 0 {
+			it.t.Fatal(errors.New("scrape 'after' parameter can't be negative or zero"))
+		}
+		if it.currTime.Add(after).After(time.Now()) {
+			it.t.Fatal(errors.New("sum of all scrape 'after' parameters can be beyond 10 hours"))
+		}
+		it.currTime = it.currTime.Add(after)
+
+		mfs, err := r.Gather()
+		if err != nil {
+			it.t.Fatal(err)
+		}
+
+		// Inject a fixed timestamps as a way to backfill/inject samples with
+		// larger (e.g. 30 seconds) intervals without waiting.
+		for _, mf := range mfs {
+			for _, m := range mf.GetMetric() {
+				m.TimestampMs = proto.Int64(timestamp.FromTime(it.currTime))
+			}
+		}
+		scrapeRecordings = append(scrapeRecordings, mfs)
+
+		return &ingestionTestExpRecorder{it: it, currTime: it.currTime}
+	})
+
+	// Once recorded, let's inject those to our backends.
+	// This might take a while.
+	// TODO(bwplotka): Potential for concurrency here.
+	for _, b := range it.backends {
+		fmt.Printf("%s Injecting samples to %v backend\n", it.t.Name(), b.b.Ref())
+		b.b.injectScrapes(it.t, scrapeRecordings, 2*time.Minute)
+	}
+}
+
+type ingestionTestExpRecorder struct {
+	it       *ingestionTest
+	currTime time.Time
+}
+
+func (ir *ingestionTestExpRecorder) Expect(val float64, metric prometheus.Metric, b Backend) ExpectationsRecorder {
+	m := dto.Metric{}
+	metric.Write(&m)
+	if m.GetSummary() != nil || m.GetHistogram() != nil {
+		// TODO(bwplotka): Implement an alternative.
+		ir.it.t.Fatal("It's not practical to use equalsGCMPromQuery against histograms and summaries.")
```

Comment:
What is `equalsGCMPromQuery`? Was that a function you had here at one point?

Reply:
yea